### ISSUE TITLE
Generalization for higher-dimensional data

### ISSUE THREAD

**Issue by awsteiner** [VincentStimper/normalizing-flows#66]
Dear normflows,

I'm having a hard time understanding something, and if you have a moment to help that would be amazing. I was looking over the real_nvp_colab example, and I wanted to make a few changes. First, I would like to use more than 2 inputs (with a different data set), and I can't quite figure out the changes necessary to make that work. My data set, labeled 'x' in the training section of the notebook, has e.g. torch.Size([6544, 3]), for ndim=3, then I change

nf.flows.Permute(2, mode='swap') to nf.flows.Permute(3, mode='swap') nf.distributions.base.DiagGaussian(2) to nf.distributions.base.DiagGaussian(3)

but then I can't quite figure out how to change

nf.nets.MLP([1, 64, 64, 2], init_zeros=True)

for the ndim=3 case. Or, is a more dramatic change to the code required for ndim=3? I get a lot of pytorch/normflows errors associated with the matrix sizes. 

Thank you,
Andrew

**Comment by VincentStimper** [comment]
Hi @awsteiner,

This is the neural net computing the parameters of a coupling flow layer. It takes in half the dimensions and returns two parameters per dimension for the other half. Since you have an odd number of dimensions, this is indeed a little bit tricky and you have to look at the code to understand what's going on.

The splitting is done [here](https://github.com/VincentStimper/normalizing-flows/blob/c6616b1a9f04c0e894b9e045e66bd504db17ef5e/normflows/flows/reshape.py#L32), using [`torch.chunk`](https://pytorch.org/docs/stable/generated/torch.chunk.html), where the first return value will have 2 dimensions and the other 1 in your case. The parameters are computed from the [first half of the dimensions](https://github.com/VincentStimper/normalizing-flows/blob/c6616b1a9f04c0e894b9e045e66bd504db17ef5e/normflows/flows/affine/coupling.py#L127), so the parameter network should expect input dimension 2. There will be only one dimension to compute parameters for, so we need output dimension 1 * 2 = 2. So the parameter network has to be `nf.nets.MLP([2, 64, 64, 2], init_zeros=True)`.

Best regards,
Vincent
